client/go/cmd/root.go

```diff
@@ -39,63 +102,77 @@ Vespa documentation: https://docs.vespa.ai`,
 		SilenceUsage:      false,
-		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
-			return configureOutput()
-		},
-		Args: cobra.MinimumNArgs(1),
+		Args:              cobra.MinimumNArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return fmt.Errorf("invalid command: %s", args[0])
 		},
 	}
+	env := make(map[string]string)
+	for _, entry := range environment {
+		parts := strings.SplitN(entry, "=", 2)
+		env[parts[0]] = parts[1]
+	}
+	version, err := version.Parse(build.Version)
+	if err != nil {
+		return nil, err
+	}
+	cli := CLI{
+		Environment: env,
+		Stdin:       os.Stdin,
+		Stdout:      stdout,
+		Stderr:      stderr,
 
-	targetArg      string
-	applicationArg string
-	waitSecsArg    int
-	colorArg       string
-	quietArg       bool
-	apiKeyFileArg  string
-	stdin          io.ReadWriter = os.Stdin
-
-	stdout = colorable.NewColorableStdout()
-	stderr = colorable.NewColorableStderr()
-)
-
-const (
-	applicationFlag = "application"
-	targetFlag      = "target"
-	waitFlag        = "wait"
-	colorFlag       = "color"
-	quietFlag       = "quiet"
-	apiKeyFileFlag  = "api-key-file"
-	apiKeyFlag      = "api-key"
-)
-
-func isTerminal() bool {
-	if f, ok := stdout.(*os.File); ok {
-		return isatty.IsTerminal(f.Fd())
+		version:    version,
+		cmd:        cmd,
+		httpClient: util.CreateClient(time.Second * 10),
+		exec:       &execSubprocess{},
 	}
-	if f, ok := stderr.(*os.File); ok {
-		return isatty.IsTerminal(f.Fd())
+	cli.isTerminal = func() bool { return isTerminal(cli.Stdout) || isTerminal(cli.Stderr) }
+	cli.configureFlags()
+	if err := cli.loadConfig(); err != nil {
+		return nil, err
 	}
-	return false
+	cli.configureCommands()
+	cmd.PersistentPreRunE = cli.configureOutput
+	return &cli, nil
 }
 
-func configureOutput() error {
-	config, err := LoadConfig()
+func (c *CLI) loadConfig() error {
+	bindings := NewConfigBindings()
+	bindings.bindFlag(targetFlag, c.cmd)
+	bindings.bindFlag(applicationFlag, c.cmd)
+	bindings.bindFlag(waitFlag, c.cmd)
+	bindings.bindFlag(colorFlag, c.cmd)
+	bindings.bindFlag(quietFlag, c.cmd)
+	bindings.bindFlag(apiKeyFileFlag, c.cmd)
+	bindings.bindEnvironment(apiKeyFlag, "VESPA_CLI_API_KEY")
+	bindings.bindEnvironment(apiKeyFileFlag, "VESPA_CLI_API_KEY_FILE")
+	config, err := loadConfig(c.Environment, bindings)
 	if err != nil {
 		return err
 	}
-	if quiet, _ := config.Get(quietFlag); quiet == "true" {
-		stdout = ioutil.Discard
+	c.config = config
+	return nil
+}
+
+func (c *CLI) configureOutput(cmd *cobra.Command, args []string) error {
+	if f, ok := c.Stdout.(*os.File); ok {
+		c.Stdout = colorable.NewColorable(f)
+	}
+	if f, ok := c.Stderr.(*os.File); ok {
+		c.Stderr = colorable.NewColorable(f)
+	}
+	if quiet, _ := c.config.get(quietFlag); quiet == "true" {
+		c.Stdout = ioutil.Discard
 	}
 	log.SetFlags(0) // No timestamps
-	log.SetOutput(stdout)
-	colorValue, _ := config.Get(colorFlag)
+	log.SetOutput(c.Stdout)
```

Comment:
There is a good mix of `log.Printf` and `fmt.Fprintf` usage, when do we use what?

Reply:
`log` should be preferred since that always point to `c.Stdout`. When using `fmt`, the caller needs to remember to pass the custom writer.

I didn't intend to update the use of `fmt` in this PR.

Reply:
My intent was only to know what to prefer in the future. Not that all fmt callers should be changed.

Thanks.